ToDo REST API 만들기

alex-gibeom opened the pull request:
안녕하세요 트레이너님
8주간 열심히 따라가보겠습니다!
잘 부탁 드립니다:)

현재 목록조회, 상세조회, 생성 API 까지만 완료하였습니다.
리팩토링은 하나도 진행하지 못해서 보기 불편하실 거 같아서 미리 죄송한 말씀 드립니다.
적나라한 피드백 리뷰 부탁드립니다.

alex-gibeom committed:
feat: Add HTTP server

alex-gibeom committed:
feat: Add ToDo 목록 조회 API

alex-gibeom committed:
feat: Add ToDo 상세 조회 API

alex-gibeom committed:
feat: Add ToDo 생성 API

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+
+public class AppHttpHandler implements HttpHandler {
+    static Long id = 0L;
+    List<Task> tasks = new ArrayList<>();
+    ObjectMapper objectMapper = new ObjectMapper();
```
이런 것들은 재할당을 방지하는 것을 고려해 봅시다.

```suggestion
    final List<Task> tasks = new ArrayList<>();
    final ObjectMapper objectMapper = new ObjectMapper();
```

alex-gibeom replied:
- objectMapper는 JsonUtil에서 클래스 로딩 시 딱 한번 생성된 변수를 재사용하도록 변경하였습니다!
- List<Task> tasks 변수는 TaskService 인스턴스가 생성될 때 초기화되도록 수정하였고 final 키워드를 달아주었습니다 (Effectively final)
   - 현재 전체 소스에서 TaskService 인스턴스는 AppHttpHandler에서만 딱 한번 초기화가 되므로 위와 같이 구현해보았습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+
+import java.io.*;
```
코드숨 과정을 진행하는 동안에는 `import *`의 사용을 지양하도록 합시다. 가급적 어떤 것을 import 하는지 명시해 줄 수 있도록 해봅시다.

alex-gibeom replied:
intellij 설정에서 import문 *을 자동으로 변경시키는 설정을 수정하였습니다 :)

참고문서: https://dev-kani.tistory.com/39

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+    ObjectMapper objectMapper = new ObjectMapper();
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
```
handle 메소드의 크기가 너무 커서 한 눈에 파악하기 어렵습니다. 각 처리 코드를 private 메소드로 추출하고 handle 메소드는 처리 로직을 읽기 좋은 형태로 수정해 보세요.

alex-gibeom replied:
데이터 처리를 담당하는 taskService 객체를 생성하여 Handle 메서드에서는 요청과 응답의 역할만 구현하도록 변경해보았습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+            userId = 0L;
+        }
+
+        if (!path.contains("/tasks")) {
+            exchange.sendResponseHeaders(404, 0);
+            responseBody.close();
+            return;
+        }
+
+        if (method.equals("PUT") || method.equals("PATCH") || method.equals("DELETE")) {
+            if (userId.equals(0)) {
+                exchange.sendResponseHeaders(400, 0);
+                responseBody.close();
+                return;
+            }
+        }
+
+        if (method.equals("POST") || method.equals("PUT") || method.equals("PATCH")) {
+            if (requestBody.isBlank()) {
+                exchange.sendResponseHeaders(400, 0);
+                responseBody.close();
```
PUT, POST, PATCH, DELETE 같은 것들은 HTTP 프로토콜에서 중요한 의미를 갖는 고정값들이니 enum으로 한번 정의해 보세요.

alex-gibeom replied:
Http Method를 enum으로 관리하도록 변경해보았습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,128 @@
+                break;
+            case "POST":
+                Task task = toTask(requestBody);
+                task.setId(++id);
```
`++id` 의 책임과 의미를 고려해 별도의 메소드나 클래스로 분리해 보세요.

alex-gibeom replied:
#### 수정 (37e31ca)
- `id`값만 관리하는 TaskId 객체를 생성하고 private 생성자를 통해 인스턴스 생성을 제한하였습니다.
   - 인스턴스 멤버가 없기 때문에 불필요한 행위 제한
- id를 증가시키고 가져오는 getNewId() 메서드를 synchronized를 통해 멀티 쓰레드 환경에 대비하였습니다.

#### 질문
- 더 나은 방법이 있는지 궁금합니다!
- synchronized를 사용하는 방법이 적절한지 궁금합니다!

johngrib replied:
이번 주차에는 이 정도로도 충분한 것 같습니다. 잘 해내셨습니다.

alex-gibeom replied:
감사합니다 :)

johngrib reviewed (commented):
반갑습니다 :-)

alex-gibeom committed:
feat: Add ToDo 제목 수정하기 API

alex-gibeom committed:
feat: Add ToDo 삭제하기 API

alex-gibeom committed:
feat: Add JsonUtil

alex-gibeom committed:
refactor: jackson 관련 리펙토링

alex-gibeom committed:
refactor: Http Method enum 타입으로 관리하도록 수정

alex-gibeom committed:
refactor: taskService 추가 후 handle service 역할 분리

alex-gibeom committed:
merge: ToDo REST API 리펙토링

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.stream.Collectors;
+
+public class AppHttpHandler implements HttpHandler {
+    TaskService taskService = new TaskService();
```
taskService가 재할당되면 곤란하겠죠.

```suggestion
    private final TaskService taskService = new TaskService();
```

alex-gibeom replied:
#### 수정 (c463ec2)
- private인 이유
   - taskService는 AppHttpHandler 내부에서만 사용되는 멤버 변수이기 때문입니다.
- final인 이유
   - taskService가 재할당 될 시 TaskService 인스턴스 변수인 tasks (데이터 목록)가 새로 초기화되기 때문에, 기존 데이터 목록이 날라가는 상황을 방지하기 위해서 final 키워드를 사용해야 합니다.

#### 잘못된 내용이 있으면 피드백 부탁드립니다!

johngrib replied:
꼼꼼하게 답변을 작성하셨네요. 열심히 학습하시는 모습이 멋집니다. 한편 이렇게 정리할 때마다 JLS 문서도 한번씩 읽어두는 습관을 길러보세요.

https://docs.oracle.com/javase/specs/jls/se11/html/jls-6.html#jls-6.6

이렇게 뭔가 찾아볼 일이 있을 때마다 공식 문서를 같이 읽는 것을 10년간 했다고 생각해 봅시다.

alex-gibeom replied:
영어가 어렵지만 공식 문서를 읽는 습관을 들여보겠습니다 :)
감사합니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,109 @@
+                || method == HttpMethodType.DELETE;
+    }
+
+    private Long getUserId(String[] pathArr) {
+        Long userId;
+        if (pathArr.length > 2) {
+            userId = Long.parseLong(pathArr[2]);
+        } else {
+            userId = 0L;
+        }
+        return userId;
+    }
```
이 메소드에 early return 방식을 적용해 보세요.

alex-gibeom replied:
#### 수정 (213aa37)
- early return 방식을 적용하였습니다.
   - if 조건절에서 값을 return 하는 방식으로 else  및 switch/case 사용을 지양하는 방식
- 장점
    - 코드의 depth가 줄어드는 장점
    - 가독성 상승

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,109 @@
+        return userId;
+    }
+
+    private boolean isEmptyUserId(Long id) {
+        return id.equals(0L);
+    }
```
`id`에 `null`이 들어오면 어떻게 될까요?

alex-gibeom replied:
id가 0으로 들어올 때의 상황을 고려하지 못하고 pathVariable이 없으면 id에 0L을 대입하여 해당 코드에서 0L로 비교하도록 이상한 행동을 했었네요...

수정하였습니다! (e4139f8)

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/HttpMethodType.java:
```diff
@@ -0,0 +1,15 @@
+package com.codesoom.assignment.enums;
+
+public enum HttpMethodType {
```
이런 경우에는 이렇게 RFC 문서도 함께 읽을 수 있도록 링크해 줍시다.
```suggestion
/**
 * @see https://www.rfc-editor.org/rfc/rfc7231
 */
public enum HttpMethodType {
```

alex-gibeom replied:
앞으로 관련 문서를 주석으로 달아보겠습니다! 감사합니다! (a93872a5)

johngrib replied:
RFC 7231 문서는 HTTP 스펙이므로 백엔드 개발자라면 개요라도 파악해둬야 하는 중요한 문서 중 하나입니다. 시간이 나실 때마다 꼭 군데군데 읽어보시고, 제대로 읽고 싶다면 "HTTP 완벽 가이드" 책을 구해서 공부해 보세요.

alex-gibeom replied:
오 안그래도 최근에 HTTP 완벽 가이드를 구입했었는데 빨리 읽어봐야겠습니다
감사합니다 :)

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,45 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
if 문을 작성할 때 "절대로" 중괄호를 생략하지 않도록 연습해보세요. 절대로 생략하지 않도록 합시다. 경력이 쌓이고 실력이 향상된 이후에도 마찬가지입니다.

이와 관련된 유명한 버그 이야기가 있습니다. Apple의 SSL/TLS 프로토콜에서 발견된 실수입니다. if 중괄호가 습관화되어 있었다면 하지 않았을 실수죠.

https://embeddedgurus.com/barr-code/2014/03/apples-gotofail-ssl-security-bug-was-easily-preventable/

alex-gibeom replied:
중괄호를 생략하지 않도록 연습하겠습니다!

intellij 설정에서 코드 자동 완성 시 Force braces 설정을 추가하였습니다.
참고 자료: https://ryumodrn.tistory.com/67

johngrib replied:
좋습니다! 이렇게 IDE를 조금씩 설정해가시면 언젠간 손에 잘 익은 자신만의 도구가 될 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/service/TaskService.java:
```diff
@@ -0,0 +1,70 @@
+         tasks = new ArrayList<>();
+    }
+
+    public String getTasks() throws IOException {
+        String data = JsonUtil.writeValue(tasks);
+
+        return data;
+    }
```
이 메소드가 `String`이 아니라 `Collection<Task>` 또는 `List<Task>`를 리턴하도록 작성해 보세요. 직접 지어준 이 클래스의 이름이 왜 `TaskService`인지, 이 이름에 맞는 코드를 어떻게 디자인 할 것인지에 대해서도 고민해 보세요.

alex-gibeom replied:
#### 수정 (bf72767)
- TaskService를 요청 데이터를 통해 필요한 응답값을 반환하는 역할만 담당하도록 변경하였습니다.
   - TaskService는 기존 AppHttpHandler가 요청 및 응답 값 관리에만 집중할 수 있도록 관심사를 분리하기 위해 생성한 객체입니다
   - 따라서 TaskService는 요청 데이터를 통해 응답에 필요한 데이터 처리를 전담하여 결과값을 반환해주는 객체로 사용하기 위해 생성하였습니다.

- Service 코드의 일관성을 유지하도록 변경하였습니다
   - 목록 조회 API 반환 타입 변경 (List<Task>)
   - 등록, 수정 반환 타입 변경 (Task)
   - 삭제 반환 타입 변경 (void)

#### 잘못된 내용이 있으면 피드백 부탁드립니다!

johngrib replied:
삭제 메소드는 삭제에 성공했는지 실패했는지를 알려주는 결과를 리턴하도록 작업해 보세요.

alex-gibeom replied:
boolean 타입으로 결과를 리턴하도록 수정완료하였습니다. (b285598)

한가지 궁금한점은 아래 사진과 같이 deleteTask()의 결과를 리턴받아서 어떤 추가 작업을 진행해야되나요?
deleteTask()를 수행하기 전에 미리 삭제할 요소가 없다면 에러 처리를 진행하고 있습니다.

![image](https://user-images.githubusercontent.com/59248326/195480808-6a3e8460-65aa-461f-a646-700b343e27d6.png)

반환값을 통해 오류 처리를 500으로 한번 더 진행 해야되는 것인지 궁금합니다

johngrib replied:
deleteTask가 실패할 수도 있다는 걸 잊으면 안 되죠. 지금은 deleteTask를 직접 작성하셨기 때문에 이 메소드가 하는 일이 뻔하다고 생각하고 계십니다. 그런데 만약 저 deleteTask를 어떤 유명한 회사에서 만들었고, 열어봤더니 코드가 백만줄이 있다고 상상해 보세요. 그리고 사용해보니 deleteTask가 작동할 때가 있고 작동하지 않을 때가 있었다고도 생각해 보세요. (알고 보니 다른 사람이 조회하고 있는 중일 때에는 삭제가 안 되는 정책이 있어서 삭제가 실패했다고도 생각해 봅시다.)

deleteTask 내부를 모른다고 생각하고 deleteTask의 리턴값을 기준으로 상식적인 프로세스를 설계해 보세요.

alex-gibeom replied:
아 넵 멘토님의 리뷰에서 deleteTask의 **삭제 성공/실패 여부를 핸들링** 하기 위해 boolean 타입으로 수정하는 것이 바람직하다는 것을 인지하고 있었습니다!

제 질문의 전달에 약간의 오류가 있었던 것 같습니다.
앞으로는 좀 더 명확하게 질문하는 연습을 해보도록 하겠습니다!

본래의 질문 의도는 deleteTask가 실패할 시 **삭제 실패 반환 오류코드**를 500으로 해야 할지, 404로 해야 할지, 기타 코드가 있을지 궁금한 것이었습니다.
(어제 찾아보니 500코드는 진짜 오류가 아닐 시 함부로 던지지 말고, 4xx 대로 핸들링 하는 것이 좋다고 하던데 마땅한 코드를 찾지 못했습니다ㅠ)

johngrib replied:
404와 500의 의미를 아직 명확하게 알지 못하셔서 생긴 의문 같습니다. 디스코드에서도 질문을 하셨고 영환님이 좋은 답변을 주셨더군요. 영환님 답변으로도 내용은 충분할 겁니다. 그런데 여기에서 멈추면 안됩니다. 또 봐야 하는 게 있어요. RFC 7231에서 404 와 500을 찾아 읽어보시고 내용을 정리해서 글을 작성해 보세요.

https://www.rfc-editor.org/rfc/rfc7231

alex-gibeom replied:
[HTTP 응답코드 공부 정리글](https://github.com/All-Rounder-Storage/programmers_knowledge_storage/blob/main/Computer-Science/%EB%84%A4%ED%8A%B8%EC%9B%8C%ED%81%AC/HTTP%20%EC%9D%91%EB%8B%B5%EC%BD%94%EB%93%9C/Alex/HTTP%20%EC%9D%91%EB%8B%B5%EC%BD%94%EB%93%9C.md)
이전에 제가 공부하면서 정리했던 응답코드를 다시 한번 보았습니다!
해당 문제에서 헷갈렸던 부분은 delete가 삭제가 실패하는 경우가 어떤 것이 있을까에서 시작되었던 것 같습니다.
평소에는 보통 서버 오류일 시 당연하게  exception handler에서 500을 반환해주기 때문에 헷갈렸던 것 같습니다.
이번 수정에서는 멘토님이 추가 정책(다른 사람이 조회하고 있는 중일 때에는 삭제가 안 되는 정책)을 주신 것에 맞게 핸들링하여, 409 conflict로 핸들링해보았습니다.

(7a9aaa4)
만약 서버의 오류로 인한 실패였다면 exception 핸들링 부분에서 500으로 반환하겠지만, 이번 주차 미션에서는 오류 핸들링을 제외하고 흐름을 제어하는 의도가 있는 것 같아서 정책으로 인한 실패로 **409 코드**로 핸들링을 해보았습니다 :)

johngrib replied:
와 열심히 공부하셨네요. 제가 다 뿌듯합니다!

alex-gibeom replied:
감사합니다ㅎㅎ!

johngrib commented on app/src/main/java/com/codesoom/assignment/service/TaskService.java:
```diff
@@ -0,0 +1,70 @@
+        return data;
+    }
+
+    public String getTaskByUserId(Long userId) throws IOException {
```
가능한 한 문제를 해결가능한 범위 내에서 가장 좁은 범위의 타입을 사용해보세요. String은 표현 범위가 너무 넓습니다. 여기에서는 `Task` 타입을 리턴하도록 수정해 보세요.

alex-gibeom replied:
#### 수정 (b3fb553)
- 상세 조회 service 반환 타입 변경 (Optional<Task>)
- 상세 조회 데이터 없을 경우 404 반환

alex-gibeom committed:
refactor: TaskId 객체를 통한 id값 관리 (관심사 분리)

Ref: #128#discussion_r991214500

alex-gibeom committed:
refactor: 인스턴스 멤버 변수 캡슐화 및 상수 처리

Ref: #128#discussion_r992261163

alex-gibeom committed:
refactor: early return 적용

Ref: #128#discussion_r992262658

alex-gibeom committed:
refactor: http 스팩 문서 링크 추가 (RFC 7231)

Ref: #128#discussion_r992263796

alex-gibeom committed:
refactor: Todo 상세 조회 API 수정

- 상세 조회 service 반환 타입 변경
- 상세 조회 데이터 없을 경우 404 반환

Ref: #128#discussion_r992269108

alex-gibeom committed:
refactor: Todo 목록 조회 API 수정 및 TaskService 일관성 작업

- 목록 조회 API 반환 타입 변경 (List<Task>)
- 등록, 수정 반환 타입 변경 (Task)
- 삭제 반환 타입 변경 (void)

Ref: #128#discussion_r992268405

alex-gibeom committed:
fix: isEmptyUserId() 메서드 내부 조건식을 0에서 null로 비교하도록 수정

- id가 0으로 들어올 때의 상황을 고려하지 못했음

Ref: #128#discussion_r992263021

alex-gibeom committed:
fix: ToDo 삭제 성공 반환코드 200에서 204로 수정

alex-gibeom committed:
merge: [Day3] - ToDo REST API 리펙토링

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/models/Task.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.Objects;
+
+public class Task {
```
Task 객체에 직렬화를 적용해야 하는건지 궁금합니다!

johngrib replied:
제가 알려드리는 것은 별로 중요하지 않습니다. 스스로 장단점을 비교해서 의사결정을 내리는 연습을 해보도록 합시다.
Task 객체에 직렬화를 적용해야 한다면 무엇 때문인지를 먼저 글로 작성해 보세요. 글로 작성하면서 장점을 목록으로 적고, 그 다음에 단점도 목록으로 적어보세요. 그리고 마지막에 둘을 종합해서 어느 쪽을 선택할지를 결론으로 적어보세요. 물론 글도 공유해 주세요.

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/error/ClientError.java:
```diff
@@ -0,0 +1,21 @@
+    public static final void badRequest(HttpExchange exchange) throws IOException {
+        exchange.sendResponseHeaders(HttpURLConnection.HTTP_BAD_REQUEST, 0);
+        exchange.getResponseBody().close();
+        throw new RuntimeException();
```
해당 요청을 바로 종료하기 위해서 RuntimeException을 발생시켰는데 해당 방법보다 좋은 방법이 있는 궁금합니다!

johngrib replied:
여기에서는 예외를 사용하지 않고 흐름 제어를 해보는 것을 연습해 봅시다. 추가 과제라고 생각해 주세요.

alex-gibeom replied:
(350ab66) 예외를 사용하지 않고 return으로 흐름 제어를 하기 위해, 기존 단일 Handler를 FrontController와 TaskController로 분리하여 역할을 나눠보았습니다!

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+            } else {
+                Optional<Task> taskByUserId = taskService.getTaskByUserId(userId);
+
+                if (taskByUserId.isEmpty()) {
+                    ClientError.notFound(exchange);
+                }
+
+                setSuccessResponse(exchange, HttpURLConnection.HTTP_OK, JsonUtil.writeValue(taskByUserId.get()));
+            }
+            return;
+        }
+
+        if (method == HttpMethodType.POST) {
+            Task task = taskService.createTask(requestBody);
+            setSuccessResponse(exchange, HttpURLConnection.HTTP_CREATED, JsonUtil.writeValue(task));
+            return;
+        }
+
+        if (method == HttpMethodType.PUT || method == HttpMethodType.PATCH) {
+            Optional<Task> taskByUserId = taskService.getTaskByUserId(userId);
+
+            if (taskByUserId.isEmpty()) {
+                ClientError.notFound(exchange);
+            }
+
+            Task task = taskService.updateTask(taskByUserId.get(), requestBody);
+            setSuccessResponse(exchange, HttpURLConnection.HTTP_OK, JsonUtil.writeValue(task));
+            return;
+        }
+
+        if (method == HttpMethodType.DELETE) {
+            Optional<Task> taskByUserId = taskService.getTaskByUserId(userId);
+
+            if (taskByUserId.isEmpty()) {
+                ClientError.notFound(exchange);
+            }
+
+            taskService.deleteTask(taskByUserId.get());
+            setSuccessResponse(exchange, HttpURLConnection.HTTP_NO_CONTENT);
+        }
```
가독성을 높이기 위해 else문과 switch/case문을 지양하고자 early return 방식을 사용한다고 하여 기존 switch/case 문을 if문으로 변경해보았는데 적절한 수정이였는지 여쭤보고싶습니다!

개인적인 생각으로 해당 메서드에서는 switch/case문이 제 눈에는 더 잘 보이는것 같아서 궁금합니다

johngrib replied:
switch를 사용하는 것도 괜찮은 방법일 것 같습니다. 그런데 중첩된 if 와 너무 긴 코드가 가독성을 꽤 해치고 있는 상황 같네요. 먼저 분기별로 짧은 요약문을 작성하는 것처럼 private 메소드들을 만들어서 보기 좋게 수정해 보세요.

alex-gibeom replied:
(350ab66) switch로 변경 후 각 역할에 맞는 private 메서드를 분리해보았습니다 (gets, get, create, update, delete)

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+                .lines()
+                .collect(Collectors.joining("\n"));
+
+        if (!path.equals("tasks")) {
```
이렇게 하면 `path`가 null이어도 NPE가 던져지지 않을 거에요.
```suggestion
        if (!"tasks".equals(path)) {
```

alex-gibeom replied:
null 위험요소를 뒤로 보내서 NPE를 피하는 방법이군요!
감사합니다 :)
해당 부분은 enum 클래스인 ContextContainer에서 stream orElse로 수정해보았습니다! (350ab66)

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+                .collect(Collectors.joining("\n"));
+
+        if (!path.equals("tasks")) {
+            ClientError.notFound(exchange);
```
이 `notFound` 메소드 안쪽에서 예외를 던지는 방식으로 흐름 제어를 하고 있을 것 같군요. 이렇게 하면 `notFound`의 구현을 모르는 사람이 `handle` 메소드를 읽을 때 무척 혼란스럽게 됩니다. 그렇게 예외를 핵심으로 제어를 하지 않는 것이 좋습니다. 만약 그 방법을 그대로 사용하고 싶다면 차라리 이렇게 return을 눈에 보이게 작성해주는 쪽이 handle 메소드를 읽는 사람에게 도움이 될 거에요.

```suggestion
            ClientError.notFound(exchange);
            return;
```

alex-gibeom replied:
(350ab66) 예외를 던지는 방식에서 직관적으로 보이도록 return으로 수정하였습니다!
감사합니다 :)

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+                || method == HttpMethodType.DELETE;
+    }
+
+    private Long getUserId(String[] pathArr) {
```
추가 과제입니다. `String[]` 타입을 사용하지 말고, `Path` 타입을 만들어서 이 타입 하나만 받도록 작업해 보세요.

alex-gibeom replied:
Path 객체를 만들어서 아래와 같이 처리되도록 수정해보았습니다!

![image](https://user-images.githubusercontent.com/59248326/195574402-b6935a9d-6524-470f-ba3c-f06f73e6d756.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/AppHttpHandler.java:
```diff
@@ -0,0 +1,131 @@
+        return pathArr.length <= 2;
+    }
+
+    private boolean isEmptyUserId(Long id) {
```
만약 요구사항이 바뀌어서 id 가 숫자가 아니라 알파벳도 포함하도록 변경된다면 어떻게 하시겠습니까?

giibeom replied:
해당 부분은 Path 객체의 pathVariable 값을 확인하는 것으로 변경하였습니다!

![image](https://user-images.githubusercontent.com/59248326/195590578-13664725-dcf2-49a7-b8c3-b084d4670290.png)

alex-gibeom replied:
(f9703fe) 이번 수정에서 id의 요구 사항이 변경될 때의 문제점을 대응하였습니다!

id의 요구사항이 추가될 시, path 내부의 isInvalidPathVariable()에 조건만 추가하면 될 것으로 보입니다

johngrib commented on app/src/main/java/com/codesoom/assignment/service/TaskService.java:
```diff
@@ -0,0 +1,49 @@
+        return task;
+    }
+
+    public Task updateTask(Task originTask, String requestBody) throws IOException {
```
`updateTask` 메소드의 입력 타입들을 다시 디자인해봅시다. 지금은 `requestBody`를 입력받고 있습니다. 그런데 만약 String이 아니라 다른 Task를 업데이트 소스로 삼으려 한다면 이 메소드를 사용할 수 있을까요? 이 메소드 외에 다른 메소드가 없다면 다른 Task를 requestBody 스트링으로 변환한다음 이 메소드에 입력해줘야 할까요?

alex-gibeom replied:
(cb7d7d4) 일종의 핸들러 매퍼 역할을 하는 TaskController에서 RequestDTO를 생성하여 reqeustBody값을 실제 컨트롤러 메서드로 넘겨주도록 수정해보았습니다!

alex-gibeom committed:
refactor: deleteTask 메서드 반환타입 삭제 성공 여부로 수정

alex-gibeom committed:
refactor: FrontController와 TaskController 역할 분리

- FrontController 역할
   - HTTP Request 데이터 세팅
   - 요청 path에 맞는 Controller로 보내주기
- TaskController 역할
   - HTTP Request 데이터 validation
   - 각 HTTP method 요청에 맞는 함수로 요청 전달
- TaskController 내부의 각 요청 method 함수들의 역할 (gets, get, create, update, delete)
   - 요청에 맞는 데이터 처리 후 Response 데이터 반환

alex-gibeom committed:
refactor: JsonUtil 메서드 제어자 순서 변경

alex-gibeom committed:
feat: Task RequestDTO 추가

alex-gibeom committed:
merge: [Day4] - ToDo REST API 리펙토링

alex-gibeom commented:
(350ab66) 아래와 같이 구조를 변경해보았습니다!
부족한 부분이나 보완할 부분이 있다면 피드백 부탁드립니다!!

![image](https://user-images.githubusercontent.com/59248326/195590000-57d5a65a-a84a-4d9b-9ca9-0521a850aa06.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Path.java:
```diff
@@ -0,0 +1,22 @@
+package com.codesoom.assignment.models;
+
+public class Path {
+    String path;
+    String pathVariable;
+
+    public String getPath() {
+        return path;
+    }
+
+    public void setPath(String path) {
+        this.path = path;
+    }
+
+    public String getPathVariable() {
+        return pathVariable;
+    }
+
+    public void setPathVariable(String pathVariable) {
+        this.pathVariable = pathVariable;
+    }
+}
```
지금 이 Path는 값만 갖고 있을 뿐 스스로 생각하거나 무언가를 결정하지 못합니다. 데이터만 갖고 있을 뿐 극도로 수동적인 객체라 할 수 있습니다. Path 바깥에 있는 Path와 관련된 코드를 여기로 응집시켜보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/FrontController.java:
```diff
@@ -0,0 +1,58 @@
+        return path;
+    }
+
+    private boolean hasPathVariable(String[] pathArr) {
+        return pathArr.length > 2;
+    }
+
+    private boolean hasPath(String[] pathArr) {
+        return pathArr.length > 0;
+    }
```
이런 검증 코드들이 `Path` 안쪽에 은닉되도록 해보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,140 @@
+                } else {
+                    get(exchange, path);
+                }
+                break;
+            case POST:
+                RequestTaskDTO.Create requestCreateDTO = JsonUtil.readValue(requestBody, RequestTaskDTO.Create.class);
+                create(exchange, requestCreateDTO);
+                break;
+            case PUT:
+            case PATCH:
+                RequestTaskDTO.Update reqeustUpdateDTO = JsonUtil.readValue(requestBody, RequestTaskDTO.Update.class);
+                update(exchange, path, reqeustUpdateDTO);
+                break;
+            case DELETE:
+                delete(exchange, path);
+                break;
+            default:
+                ClientError.methodNotAllowed(exchange);
```
여기에서는 break보다 더 강력한 return 을 사용해보세요.

alex-gibeom replied:
수정하였습니다 :)

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/ContextContainer.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.Arrays;
+
+public enum ContextContainer {
```
이 enum은 무엇을 책임지고 있나요?

alex-gibeom replied:
해당 enum을 생성한 계기는 일종의 bean 관리를 하나의 객체에서 담당하도록 해보고 싶어서 생성해보았습니다.
ContextContainer의 역할은 context (등록되어 있는 path)를 관리하며 해당 path에 맞는 enum 값을 반환해주는 역할입니다.
이 후 해당 반환 타입을 가지고 FrontController로 넘어가서 해당 값에 맞는 컨트롤러(TaskController)로 연결해줍니다.

johngrib replied:
좋습니다. JavaDoc 작성방법을 조사하시고, 설명하신 내용을 JavaDoc으로 `ContextContainer`에 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/HttpMethodType.java:
```diff
@@ -0,0 +1,39 @@
+                .orElse(HttpMethodType.NOT_SUPPORT);
+    }
+
+    // 위와 아래중에 무엇이 더 나은지 궁금합니다!
+    public static HttpMethodType getMethod1(String method) {
+        return isSupportedMethod(method)
+                ? HttpMethodType.valueOf(method)
+                : HttpMethodType.NOT_SUPPORT;
+    }
+
+    public static boolean isSupportedMethod(String method) {
+        return Arrays.stream(HttpMethodType.values())
+                .anyMatch(methodType -> methodType.getMethodValue().equals(method));
+    }
```
이건 취향을 탑니다. 제 취향으로는 이런 메소드에서는 stream을 굳이 쓰지 않고 평범한 for 를 썼을 것 같네요.

alex-gibeom replied:
아하 넵 알겠습니다!

johngrib replied:
제 취향과 똑같이 작업하지 않으셔도 됩니다! 취향을 언급한 이유이기도 합니다. 본인의 취향대로 하셔도 돼요 ㅎㅎㅎ

alex-gibeom committed:
refactor: Path 관련 플로우(프로세스) 변경

- requestMapping 메서드를 통해 아래와 같은 역할을 부여
   - request 데이터 체크
   - 본 컨트롤러 메서드에서 받는 매개변수 세팅
   - 매개변수 파라미터 타입 체크

alex-gibeom committed:
refactor: HttpMethodType 질문용 메서드 제거

alex-gibeom committed:
merge: [Day5] - ToDo REST API 리펙토링 (requestMapping)

alex-gibeom commented:
멘토님 덕분에 부족하지만 많이 배워가고 있습니다!
열심히 따라가겠습니다 감사합니다 :)

금일 수정 내용을 정리해보았습니다 (7e98242)
### Path 관련 프로세스 변경 (requestMapping 메서드)

#### [변경한 계기]
- Path는 본 컨트롤러 메서드(gets, get, create, update, delete)로 오기 전에 이미 아래와 같이 받고 싶은 형태로 변경되어 들어와야 된다고 생각하였습니다

![image](https://user-images.githubusercontent.com/59248326/195831559-4debc3d7-ce11-4e63-9f0f-bb3043c2b203.png)

- 따라서 @PathVariable 역할을 requestMapping 메서드에서 수행하도록 만들어보았습니다


#### [기존]
- PathVariable(id) 값을 본 컨트롤러 메서드 내부에서 변경하였습니다.
- Path의 검증 및 세팅이 다방면에 퍼져있었습니다.
- Path 관련 코드가 외부에 퍼져있었습니다.

#### [변경]
1. TaskController에서 본 컨트롤러 메서드로 가기 전에 reqeustMapping 역할을 모두 수행하고, 본 컨트롤러 메서드로 전달하도록 변경하였습니다.
    - requestMapping 역할
        - request 데이터 체크
        - 본 컨트롤러 메서드에서 받는 매개변수 세팅
        - 매개변수 파라미터 타입 체크
2. path 관련 작업 메서드를 모두 Path 객체로 응집하였습니다.
3. hasPathVariable 삭제
    - hasPathVariable은 순수 path의 depth가 2 이상인 경우가 있을 수 밖에 없는데, 무작정 2번째 인덱스를 pathVariable로 설정하는 것은 논리적 오류라고 판단하였습니다
    - 해당 역할은 ContextContainer에서 context를 찾는 메서드로 위임하였습니다.
4. hasPath 삭제
    - path 자체를 ContextContainer에서 비교하여, 등록되어 있는 path와 일치하는게 없다면 NOT_EXIST로 반환하도록 변경하였으므로 필요가 없어져서 삭제하였습니다.

alex-gibeom commented:
두가지의 궁금한 점을 질문드리고 싶습니다!

### 질문 1
- 애플리케이션의 흐름도를 수정, 보완해보았습니다
- 멘토님이 보셨을 때 **보완할 점**이 있을지, **좀 더 좋은 방법이 있을지** 의견을 들어보고 싶습니다!

![image](https://user-images.githubusercontent.com/59248326/195831348-54ed39a3-4afc-475a-8bc5-32b5756b4bd8.png)

<br>

### 질문 2
- 해당 중복코드를 어떻게 제거하면 좋을지 고민 끝에 답이 안나와서 질문드립니다.
- 예외처리(throws)를 사용하지 않고 흐름을 제어하려면 switch문 내에서 return을 사용해야 하는데, 
그렇게 함으로써 GET, PATCH, DELETE에서 중복코드가 발생됩니다…!

![image](https://user-images.githubusercontent.com/59248326/195832580-92de5ea0-2d39-435f-9396-0e7b6fc3e89f.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,155 @@
+            return;
+        }
+
+        switch (method) {
```
switch 블록의 크기가 너무 큽니다. switch 라는 단어의 뜻을 잘 생각해 보시고 switch가 switch의 역할만 하도록 크기를 줄여보세요. 전체 흐름을 파악하는 데에 필요한 코드는 짧고 명확할 수록 좋습니다. 반면 흐름 파악보다는 작업의 디테일을 묘사하는 것이 중요한 코드는 굳이 크기를 줄일 필요가 없겠죠.

예를 들어 다음과 같은 방식을 생각해 볼 수 있을 겁니다.

```
switch (method) {
  case GET: return responseTask(searchRequest);
  case POST: return createTask(saveRequest);
  ...
}
```

alex-gibeom replied:
(4de8180)
switch를 간략하게 간추리고,  각각의 controller 메서드를 가기 전 Advice 메서드를 거치도록 변경하였습니다.

각각의 Advice 메서드의 역할은 pathVaraible 검증과 taskId 값을 세팅 후, 실제 controller로 전달해주기 위해 생성하였습니다.

```java
switch (method) {
    case GET:
        responseTaskAdvice(exchange, path);
        return;
    case POST:
        createTaskAdvice(exchange, requestBody);
        return;
    case PUT:
    case PATCH:
        updateTaskAdvice(exchange, path, requestBody);
        return;
    case DELETE:
        deleteTaskAdvice(exchange, path);
        return;
    default:
        ClientError.methodNotAllowed(exchange);
}
```

johngrib replied:
이제 각각의 처리를 별도의 책임을 갖는 개별 클래스 파일로 분할해 봅시다.

alex-gibeom replied:
이번 경험을 통해 책임을 분리하여 각 객체의 응집도를 높이는 연습을 하게 된 것 같습니다.
이제껏 코딩을 하면서 응집도, 결합도 등을 많이 생각을 안하고 했던게 반성이 되네요.
하나의 객체에는 1개의 관심사를, 하나의 메서드에는 1개의 기능을 갖게 하도록 계속해서 신경쓰면서 코딩해야겠습니다!

(f3a3b7c)
기존 switch가 있는 클래스를 TaskControllerAdvice로 변경하고,
TaskController 클래스를 생성하여 실제 컨트롤러 메서드를 전부 이관해보았습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/enums/ContextContainer.java:
```diff
@@ -0,0 +1,58 @@
+        return NOT_EXIST;
+    }
+    
+/*      (1) 해당 조건이 만족한다는 것은 제일 끝 path가 pathVariable
+
+        해당 로직 설계 이유
+         만약 해당 경로들이 들어온다면..?
+         - /tasks/2
+         - /tasks/giibeom
+
+         현재 context 빈에 등록되어 있는 경로들 (예시)
+            /tasks              -> 얘를 반환해야됨
+            /tasks/test
+            /user/tasks/test
+
+        맨 마지막 패스를 PathVariable이라고 가정하였을 때,
+        그걸 빼고 일치하는 경로가 있으면 해당 context를 반환하도록 하였습니다
+        또한 마지막 패스가 PathVariable이 맞다는 소리이므로 PathVariable 세팅도 진행하였습니다.
+
+
+        (2)
+        call by value이지만 같은 참조값을 바라보기 때문에, 값이 정상적으로 바뀔 것이라고 생각합니다.
+*/
```
이 주석이 말하고자 하는 바와 의도가 잘 이해되지 않습니다. 이 주석은 현재 @giibeom 님에게는 이해하기 쉬워도 여러 PR을 이어서 리뷰하는 사람에게는 이해하기 어려운 글일 수 있어요. 작문을 한다고 생각하시고 생각하시는 바를 명확하게 서론/본론/결론이 있는 글로 작성해 보세요. 주석의 형태로 작성하지 않는 것이 나을 수도 있습니다. 일종의 글쓰기 훈련으로 생각하고 접근하셔도 됩니다.

alex-gibeom replied:
이걸 작성하고 다음날 제가 봐도 이건 도대체 무슨글인가... 싶었습니다.
그냥 간단하게 맨끝의 path가 pathVariable인지 검증하려는 목적이였는데 너무 횡설수설 했던 것 같습니다

해당 부분은 간단하게 startsWith 메서드를 통해 해결하였습니다!
글쓰기 공부를 좀 해봐야겠네요.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,155 @@
+                    return;
+                }
+
+                if (!StringUtil.isNumeric(path.getPathVariable())) {
```
이런 모양이 되도록 수정해 보세요.

```suggestion
                if (!path.hasTaskId()) {
```

이렇게 할 수도 있겠죠.

```suggestion
                if (!path.isRepresentsTaskIdentity()) {
```

alex-gibeom replied:
(88fba6b, f9703fe)

Path 객체에게 pathVariable을 검증하는 메서드를 생성해주며, path 검증은 Path 객체에만 의존하도록 변경하였습니다!
```java
if (!path.isInvalidPathVariable()) {
    ClientError.methodArgumentTypeMismatch(exchange);
    return;
}


// Path 객체 내부
public boolean isInvalidPathVariable() {
    return StringUtil.isNumeric(this.pathVariable);
}


// StringUtil 내부
public static boolean isNumeric(String str) {
    return str.chars()
            .allMatch(Character::isDigit);
}
```

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,155 @@
+
+        switch (method) {
+            case GET:
+                if (path.getPathVariable() == null) {
```
`null`을 쓰지 않고 이 `if`문이 일상어처럼 읽히도록 코드를 수정해 보세요.
예를 들어 이렇게 할 수 있겠죠.

```suggestion
                if (path.isInvalid()) {
```

alex-gibeom replied:
비교식도 일상어처럼 읽히도록 메서드로 추출하는 것을 습관들여야겠네요 :)

해당 조건식은 pathVariable이 비어있는지 확인하는 조건으로써, isEmptyPathVariable(path)를 사용해보았습니다!
```java
if (isEmptyPathVariable(path)) {
    getTaskList(exchange);
    return;
}

private boolean isEmptyPathVariable(Path path) {
    return path.getPathVariable() == null;
}
```

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,155 @@
+
+                return;
+            case PUT:
+            case PATCH:
+                if (!StringUtil.isNumeric(path.getPathVariable())) {
```
path가 숫자인지 아닌지는 PATCH에게는 별로 중요한 문제가 아닐 수 있습니다. 그건 path에게 맡겨봅시다.

alex-gibeom replied:
위 답변과 동일한 내용입니다!

(88fba6b, f9703fe)

Path 객체에게 pathVariable을 검증하는 메서드를 생성해주며, path 검증은 Path 객체에만 의존하도록 변경하였습니다!
```java
if (!path.isInvalidPathVariable()) {
    ClientError.methodArgumentTypeMismatch(exchange);
    return;
}


// Path 객체 내부
public boolean isInvalidPathVariable() {
    return StringUtil.isNumeric(this.pathVariable);
}
```

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringUtil.java:
```diff
@@ -0,0 +1,13 @@
+public class StringUtil {
+
+    public static boolean isNumeric(String str) {
+        try {
+            Long.parseLong(str);
+            return true;
+        } catch(NumberFormatException e) {
+            return false;
+        }
```
try/catch를 사용하지 않는 방식으로도 이 문제를 풀어봅시다.

alex-gibeom replied:
(adc801a)
해당 방식은 Java 8의 Stream API를 통해 try/catch를 사용하지 않도록 변경해보았습니다!

``` java
public static boolean isNumeric(String str) {
    return str.chars()
            .allMatch(Character::isDigit);
}
```

johngrib replied:
추가 과제입니다. `0`으로 시작하는 것을 허용하지 않아야 한다는 조건이 주어졌다고 합시다.

alex-gibeom replied:
추가 과제 감사드립니다 :)
(692bf39) 멘토님의 과제가 두가지로 해석이 되어 두개 다 구현해 보았습니다.

1. 문자열이 0보다 큰 값인지
```java
public static boolean isOverThanZero(String str) {
    if (isNull(str)) {
        return false;
    }

    if (!isNumeric(str)) {
        return false;
    }

    return Integer.parseInt(str) > 0;
}

// 숫자인지 확인 메서드
public static boolean isNumeric(String str) {
    if (isNull(str)) {
        return false;
    }

    return str.chars()
            .allMatch(Character::isDigit);
}
```

2. 문자열이 0으로 시작하는지
```java
public static boolean startWithZero(String str) {
    if (isNull(str)) {
        return false;
    }

    return str.startsWith("0");
}
```

johngrib commented:
> 두가지의 궁금한 점을 질문드리고 싶습니다!
> 
> ### 질문 1
> * 애플리케이션의 흐름도를 수정, 보완해보았습니다
> * 멘토님이 보셨을 때 **보완할 점**이 있을지, **좀 더 좋은 방법이 있을지** 의견을 들어보고 싶습니다!
> 
> ![image](https://user-images.githubusercontent.com/59248326/195831348-54ed39a3-4afc-475a-8bc5-32b5756b4bd8.png)
> 

아직까지는 문제가 단순해서 큰 문제는 없어 보입니다.

> 
> ### 질문 2
> * 해당 중복코드를 어떻게 제거하면 좋을지 고민 끝에 답이 안나와서 질문드립니다.
> * 예외처리(throws)를 사용하지 않고 흐름을 제어하려면 switch문 내에서 return을 사용해야 하는데,
>   그렇게 함으로써 GET, PATCH, DELETE에서 중복코드가 발생됩니다…!
> 
> ![image](https://user-images.githubusercontent.com/59248326/195832580-92de5ea0-2d39-435f-9396-0e7b6fc3e89f.png)

오늘 코드 리뷰에 남긴 것들을 해결하시면 이 문제도 같이 해결될 가능성이 높습니다.

alex-gibeom committed:
refactor: Path객체에 pathVariable numeric인지 검사하는 메서드 생성

- Path 관련 작업은 Path객체로 모두 응집

alex-gibeom committed:
refactor: StringUtil isNumeric 메서드 try/catch 사용 안하도록 변경

- Java 8 Stream API 사용

alex-gibeom committed:
refactor: requestMapping 내 switch문 간결화 - Controller Advice 추가

alex-gibeom committed:
refactor: Path 객체 pathVariable 검증 메서드 변경

- pathVariable이 numeric이 아니게 요구사항이 변경될 수도 있다는 점 대비

alex-gibeom committed:
fix: task 삭제 실패 핸들링 - 삭제 불가 정책 추가

- 다른 사람이 조회하고 있는 중일 때에는 삭제가 안 되는 정책

alex-gibeom committed:
merge: [Day6] - ToDo REST API 리펙토링

alex-gibeom commented:
> 오늘 코드 리뷰에 남긴 것들을 해결하시면 이 문제도 같이 해결될 가능성이 높습니다.

해당 중복 부분은 해결 실패하였습니다...
아직 많이 부족한가 봅니다...ㅠㅠ

![image](https://user-images.githubusercontent.com/59248326/195972215-ef7ad77c-aa21-400d-8191-4d56f6d173b0.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Path.java:
```diff
@@ -0,0 +1,50 @@
+import com.codesoom.assignment.utils.StringUtil;
+
+public class Path {
+    String path;
+    String pathVariable;
```
이 값들을 재할당이 방지된 비밀 멤버로 수정해 보세요. 생성자 작업이 필요할 겁니다.
```suggestion
    private final String path;
    private final String pathVariable;
```

alex-gibeom replied:
기존 path는 최초 생성 후 ContextContainer에서 다시 setPathvariable을 하도록 설계해버려서 캡슐화 및 상수처리를 못했었습니다.

따라서 설계를 변경하여 (e4a0d9b) 수정 내역에서 변수들을 private final 처리 후 매개변수 생성자를 통해 인스턴스가 생성되도록 수정해보았습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/models/Path.java:
```diff
@@ -0,0 +1,50 @@
+        return pathVariable;
+    }
+
+    private void setPath(String path) {
+        this.path = path;
+    }
+
+    private void setPathVariable(String pathVariable) {
+        this.pathVariable = pathVariable;
+    }
```
setter 가 있다면 Path의 작동을 믿지 못하게 되는 순간이 올 지도 모릅니다.
```suggestion
```

alex-gibeom replied:
앞으로 setter 사용을 지양해야겠습니다
감사합니다 :)

(e4a0d9b) setter 부분은 모두 삭제하였습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/service/TaskService.java:
```diff
@@ -0,0 +1,53 @@
+        return taskDataList;
+    }
+
+    public Optional<Task> getByUserId(Long userId) {
+        return taskDataList.stream()
+                .filter(t -> t.getId().equals(userId))
+                .findFirst();
+    }
```
사용자별로 여러 Task를 등록할 수 있게 된다면 어떻게 하시겠습니까?

alex-gibeom replied:
헉 아직 user 도메인이 없는데도 userId를 변수명으로 선언했네요..!
taskId인데 잘못한 것 같습니다...!

해당 함수는 taskId를 통해 해당 task의 상세조회를 해주는 메서드인데 착오가 있었습니다.

멘토님 질문의 대한 답변은 만약 사용자 도메인이 있다는 가정하에, 사용자별로 여러 Task를 가져오려면 findFirst()가 아닌 `.collect(Collectors.toList());` 을 통한 List<Task>를 통해 반환할 것 같습니다
```java
public List<Task> getByUserId(Long userId) {
    return taskDataList.stream()
            .filter(t -> t.getId().equals(userId))
            .collect(Collectors.toList());
}
```

(f3d96e7) 수정하였습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringUtil.java:
```diff
@@ -0,0 +1,9 @@
+
+public class StringUtil {
+
+    public static boolean isNumeric(String str) {
+        return str.chars()
+                .allMatch(Character::isDigit);
+    }
```
str이 null 이면 어떻게 될까요?

alex-gibeom replied:
NPE가 뜰 것 같습니다...ㅠ
null 체크를 항상 염두해야 되겠네요..!

(692bf39) 수정하였습니다!

alex-gibeom committed:
refactor: 클래스 생성을 통해 관심사(책임) 분리 - ControllerHandler

- 기존 TaskController는 TaskControllerAdvice 객체로 변경
   - [TaskControllerAdvice 역할]
   - request 데이터를 검증 및 DTO 객체에 담아서 실제 컨트롤러로 보내주는 역할만 담당
- TaskController 객체 생성 후 실제 Controller 메서드를 모두 이관
   - [TaskController 역할]
   - request 데이터를 통해 필요한 데이터 처리를 하는 service 호출
   - response 데이터 생성

alex-gibeom committed:
feat: pathVariable이 0으로 시작하는 것을 허용하지 않는 조건 추가

1. 문자열이 0보다 큰 값만 허용하는 경우
2. 문자열 자체가 0으로 시작하는 것을 허용하지 않는 경우

alex-gibeom committed:
fix: 모든 userId 관련 변수 및 메서드명 taskId로 변경

- user 도메인은 아직 없음

alex-gibeom committed:
refactor: Path 멤버 변수 private final 작업

- 매개변수 생성자 생성
- ContextContainer에서 하던 Pathvariable Set 작업 삭제
   - context 반환 후 해당 context 값을 통해 Path 생성 하도록 변경

alex-gibeom committed:
merge: [Day7] - ToDo REST API 리펙토링

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/models/RequestTaskDTO.java:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        public void setTitle(String title) {
+            this.title = title;
+        }
+
+        public Task toEntity() {
+            Task task = new Task();
+            task.setTitle(title);
+            return task;
+        }
+    }
+
+    public static class Update {
+        private Long id;
+        private String title;
+
+        public Long getId() {
+            return id;
+        }
+
+        public void setId(Long id) {
+            this.id = id;
+        }
+
+        public String getTitle() {
+            return title;
+        }
+
+        public void setTitle(String title) {
+            this.title = title;
+        }
+
+        public Task toEntity() {
+            Task task = new Task();
+            task.setId(id);
+            task.setTitle(title);
+            return task;
+        }
+    }
```
한가지 궁금한 점이 있습니다!
기존에는 request 데이터를 담도록 DTO 클래스를 각 CRUD마다 파일을 생성하였습니다
ex) RequestCreateTaskDTO, RequestUpdateTaskDTO ....

이번에 기존에 썼던 방식과는 다르게 DTO 클래스 내부에서, inner class를 통해 create와 update 용 객체를 생성하였는데 이런식으로도 실무에서 많이 사용되는지 궁금합니다!!

johngrib replied:
많이... 라고 할 수는 없고요, 이렇게 쓰는 경우도 있긴 있습니다.

alex-gibeom replied:
아하 그렇군요.. 평소에 궁금했던거였는데 답변 감사드립니다ㅎㅎ!

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringUtil.java:
```diff
@@ -0,0 +1,37 @@
+        return str.startsWith("0");
+    }
+
+    public static boolean isNull(String str) {
+        return str == null;
+    }
```
이 메소드는 없어도 괜찮을 것 같군요 ㅎㅎ

alex-gibeom replied:
아 간단한 null 체크까지는 일상어로 변경할 필요는 없군요!
감사합니다 :-)

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/StringUtil.java:
```diff
@@ -0,0 +1,37 @@
+
+public class StringUtil {
+
+    public static boolean isNumeric(String str) {
+        if (isNull(str)) {
+            return false;
+        }
+
+        return str.chars()
+                .allMatch(Character::isDigit);
+    }
```
이 메소드가 `0`은 허용하지만 `01`은 허용하지 않아야 한다고 생각하고 수정해 봅시다.

giibeom replied:
아하 그렇다면 startsWith 함수를 추가로 사용하면 될 것 같습니다!

```java
public static boolean isNumeric(String str) {
    if (str == null) {
        return false;
    }

    if (str.startsWith("0")) {
        return false;
    }

    return str.chars()
            .allMatch(Character::isDigit);
}
```

alex-gibeom replied:
빠르게 생각해보았을때는 0이 아니면서 0으로 시작하는지 startsWith() 함수로 확인해 볼 것 같습니다!

(eb4bb42) 수정하였습니다!
```java
public static boolean isNumeric(String str) {
    if (str == null) {
        return false;
    }

    if (!str.equals("0") && str.startsWith("0")) {
        return false;
    }

    return str.chars()
            .allMatch(Character::isDigit);
}
```

johngrib reviewed (approved):
1주차인데도 굉장히 열심히 해 주셨고, 쉽지 않은 여러 추가 과제도 잘 해내셨습니다. 주간 회고를 작성해 공유해 주세요.

alex-gibeom committed:
fix: isNumeric 함수 내부에서 0이 아닐 때 0으로 시작하는 값을 허용하지 않도록 변경

alex-gibeom commented:
> 1주차인데도 굉장히 열심히 해 주셨고, 쉽지 않은 여러 추가 과제도 잘 해내셨습니다. 주간 회고를 작성해 공유해 주세요.

한주간 질문을 너무 많이 한 것 같은데도 세심하게 리뷰해주셔서 정말 너무너무 감사드립니다 :-)
앞으로도 열심히 학습하고 따라가겠습니다
주간회고 글은 디스코드 주간회고 채널에 올렸습니다!
